worker/uniter/pebblenotices_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package uniter_test
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/canonical/pebble/client"
+	"github.com/juju/clock/testclock"
+	"github.com/juju/loggo"
+	jc "github.com/juju/testing/checkers"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/workertest"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/testing"
+	"github.com/juju/juju/worker/uniter"
+	"github.com/juju/juju/worker/uniter/container"
+)
+
+type pebbleNoticerSuite struct {
+	clock             *testclock.Clock
+	worker            worker.Worker
+	clients           map[string]*fakePebbleClient
+	workloadEventChan chan string
+	workloadEvents    container.WorkloadEvents
+}
+
+var _ = gc.Suite(&pebbleNoticerSuite{})
+
+func (s *pebbleNoticerSuite) setUpWorker(c *gc.C, containerNames []string) {
+	s.clock = testclock.NewClock(time.Time{})
+	s.workloadEventChan = make(chan string)
+	s.workloadEvents = container.NewWorkloadEvents()
+	s.clients = make(map[string]*fakePebbleClient)
+	for _, name := range containerNames {
+		s.clients[name] = &fakePebbleClient{
+			clock:       testclock.NewClock(time.Time{}),
+			noticeAdded: make(chan *client.Notice, 1), // buffered so AddNotice doesn't block
+		}
+	}
+	newClient := func(cfg *client.Config) (uniter.PebbleClient, error) {
+		c.Assert(cfg.Socket, gc.Matches, pebbleSocketPathRegexpString)
+		matches := pebbleSocketPathRegexp.FindAllStringSubmatch(cfg.Socket, 1)
+		return s.clients[matches[0][1]], nil
+	}
+	s.worker = uniter.NewPebbleNoticer(loggo.GetLogger("test"), s.clock, containerNames, s.workloadEventChan, s.workloadEvents, newClient)
+}
+
+func (s *pebbleNoticerSuite) waitWorkloadEvent(c *gc.C, expected container.WorkloadEvent) {
+	select {
+	case id := <-s.workloadEventChan:
+		event, callback, err := s.workloadEvents.GetWorkloadEvent(id)
+		c.Assert(err, jc.ErrorIsNil)
+		c.Assert(event, gc.DeepEquals, expected)
+		callback(nil)
+	case <-time.After(testing.LongWait):
+		c.Fatalf("timed out waiting for event")
+	}
+}
+
+func (s *pebbleNoticerSuite) TestWaitNotices(c *gc.C) {
```

Comment:
`go test -c && go run golang.org/x/tools/cmd/stress ./uniter.test -check.f=pebbleNoticerSuite.TestWaitNotices`

Flaky test:
```
FAIL: pebblenotices_test.go:64: pebbleNoticerSuite.TestWaitNotices

pebblenotices_test.go:114:
    ...
/home/hpidcock/go/pkg/mod/github.com/juju/worker/v3@v3.4.0/workertest/check.go:83:
    c.Check(err, jc.ErrorIsNil)
... value *errors.Err = &errors.unformatter{message:"", cause:(*errors.errorString)(0x75ce350), previous:(*errors.errorString)(0x75ce350), function:"github.com/juju/juju/worker/uniter.(*pebbleNoticer).run", line:126} ("tomb: dying")
... error stack:
	tomb: dying
	github.com/juju/juju/worker/uniter.(*pebbleNoticer).run:126:

2023-12-15 00:25:09 ERROR test pebblenotices.go:108 container "c1": WaitNotices error, waiting 1s: WaitNotices error!
OOPS: 3 passed, 1 FAILED
--- FAIL: TestPackage (0.91s)
FAIL
```

Reply:
Thanks! Per discussion, was caused by wrapping `tomb.ErrDying` with `errors.Trace` -- removed the wrapping (but maybe one day you guys can improve the tomb package to use `errors.Is` instead of equality).